docstring on __eq__ function on quanumcircuit and dagcircuit.

uchoi91 opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [ ] I have added the tests to cover my changes.
- [ ] I have updated the documentation accordingly.
- [ ] I have read the CONTRIBUTING document.
-->

### Summary
The issue was missing doc on __eq__ on both quantum and dag circuit. 
I have add docstring for both __eq__ function and examples.


### Details and comments
fix #9044

uchoi91 committed:
docstring on __eq__ function on quanumcircuit and dagcircuit.

based upon issue : 9044
I wrote docstring on eq function on quantum and dag circuit to provide what are being considered and examples associated.

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=10196) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=10196) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=10196) it.</sub>

CLAassistant commented:
[![CLA assistant check](https://cla-assistant.io/pull/badge/not_signed)](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=10196) <br/>Thank you for your submission! We really appreciate it. Like many open source projects, we ask that you sign our [Contributor License Agreement](https://cla-assistant.io/Qiskit/qiskit-terra?pullRequest=10196) before we can accept your contribution.<br/><sub>You have signed the CLA already but the status is still pending? Let us [recheck](https://cla-assistant.io/check/Qiskit/qiskit-terra?pullRequest=10196) it.</sub>

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

jlapeyre commented:
Hi @uchoi91 . Thanks for the PR. Regardless of the final form the PR takes, you'll need to sign the CLA as noted above. If you have any trouble with this, please leave a note here.

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

javabster commented:
hi @uchoi91 currently where you have put this docstring it will not render in the documentation website. I believe what was suggested in the original issue was that this text should be moved up to the class level docstring (so the docstring right at the top around line 135) not in the `__eq__` docstring itself.

uchoi91 committed:
missing doc at class level

using sphinx to add the automethod and updated the doc that contain code snippet and refer to a Operator class.

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

uchoi91 commented:
Thanks @javabster for catching that.
First time going through the doc update didn't realize there is sphinx/makefile that generates the documentation.
I have made changes (using automethod) to populate the doc at the class level.

javabster commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -504,6 +507,30 @@ def __str__(self) -> str:
 
     def __eq__(self, other) -> bool:
+        """Check if two circuits are equal.
+        This will check .cablirations and .global_phrase
```
for code references like `.calibrations` etc. these should be written using sphinx's referencing syntax (I think you might need the `:atttr:` one for these). You can find some examples of it in other docstrings, and you did it correctly already with `:class:.Operator` 😄

uchoi91 replied:
hey @javabster 
I am looking at attr and some references are having issues.
For example, in quantumcircuit, the when adding .name with attr, 
I'm getting this error locally: qiskit-terra/qiskit/circuit/quantumcircuit.py:docstring of qiskit.circuit.quantumcircuit.QuantumCircuit.__eq__:1: WARNING: more than one target found for cross-reference 'name'
so I tried adding this code below in the python file
```
@property
    def name(self) -> str:
```
and that throws error Failed to import qiskit.compiler. where hint points to most likely due to a circular import.
in this scenario, what would be the correct way to approach this?
Just testing things, I found out I could add :attr:`name` where it will treat it as attr but won't do anything when clicking on it.

Thanks in advance!

javabster replied:
hi @uchoi91 you shouldn't add a new property to the file just to fix a sphinx error. I'm not sure exactly what the issue is here but I suspect it is because you are not being specific enough when specifying the name attribute in your docstring. If you just put `:attr:.name` then sphinx doesn't know which name you are trying to link to, as many classes in the terra documentation have attributes called `name`. I think you'd need to use `:attr:QuantumCircuit.name` or something. You can check out some of the other files in the repo for examples of how they use `:attr:` in their doc strings :)

uchoi91 replied:
@javabster Thanks your patience as I got the changes pushed out with my latest commit. 
Not sure what the exact process is but marking resolved from my end since I did push out changes to your comment.

javabster commented on qiskit/dagcircuit/dagcircuit.py:
```diff
@@ -1007,6 +1009,34 @@ def num_tensor_factors(self):
+                qc1.x(0)
+                dc1 = circuit_to_dag(qc1);
+
+                DAGCircuit.__eq__(dc, dc1)
```
for code examples that you want to show an output for I recommend using the `.. testcode:` and `.. testoutput` directives. there's some explanation [here](https://qiskit.github.io/qiskit_sphinx_theme/sphinx_guide/how_to_add_code.html#testable-cells) on how to use these directives, and you can check out some examples in our current docs in `docs/migration_guides/algorithms_migration.rst`

uchoi91 replied:
Thanks for the doc links and examples! got this changed locally.

javabster reviewed (commented):
thanks for working on this @uchoi91, I left some comments but generally looking good 😄

atharva-satpute commented on qiskit/dagcircuit/dagcircuit.py:
```diff
@@ -1007,6 +1009,34 @@ def num_tensor_factors(self):
 
     def __eq__(self, other):
+        """Check if two circuits are equal.
+        This will check .cablirations and .global_phrase and
```
```suggestion
        This will check .calibrations and .global_phrase and
```

uchoi91 replied:
Thanks! and yup I saw this spelling error on both the documentation

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

uchoi91 committed:
updated using sphinx's reference syntax. using attr and testcode.

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

atharva-satpute commented on qiskit/dagcircuit/dagcircuit.py:
```diff
@@ -1010,32 +1010,35 @@ def num_tensor_factors(self):
+        and will not consider :attr:`~DAGCircuit.data`, :attr:`~DAGCircuit.name`,
+        :attr:`~DAGCircuit.metadata` when making determination.
+        The reason for this is due to being hard and computationally expensive to perform.
+        Please refer to :class:`.Operator` if you need to achieve that level of comparision
```
```suggestion
        Please refer to :class:`.Operator` if you need to achieve that level of comparison
```

uchoi91 replied:
Thanks for catching spelling err!

uchoi91 committed:
Update qiskit/dagcircuit/dagcircuit.py with spelling err

Co-authored-by: atharva-satpute <55058959+atharva-satpute@users.noreply.github.com>

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

uchoi91 committed:
Merge branch 'main' into circuit-eq-docstring

Cryoris commented on qiskit/circuit/quantumcircuit.py:
```diff
@@ -491,6 +494,34 @@ def __str__(self) -> str:
         return str(self.draw(output="text"))
 
     def __eq__(self, other) -> bool:
+        """Check if two circuits are equal.
+        This will check :attr:`~QuantumCircuit.calibrations` and :attr:`~QuantumCircuit.global_phase`
+        and will not consider :attr:`~QuantumCircuit.data`, :attr:`~QuantumCircuit.name`,
+        :attr:`~QuantumCircuit.metadata` when making determination.
+        The reason for this is due to being hard and computationally expensive to perform.
+        Please refer to :class:`.Operator` if you need to achieve that level of comparison.
```
The equality does take into account `QuantumCircuit.data` -- it doesn't require the insertion order of the gates to be the same, since the circuit implements the same unitary still 🙂 The actual comparison is implemented on the level of the DAG representation, you can check the implementation here; https://github.com/Qiskit/qiskit/blob/4c1de131201164d5a2116342b9a431b34a92d917/crates/circuit/src/dag_circuit.rs#L2404.

If you still would like to contribute, we should update this to state the the DAG representation of both circuits, as well as the qubits, registers and metadata is compared. I like your note about the operator, we could add this stating that "to compare whether circuits with different gate representations implement the same unitary, refer to the Operator class".
